### PR TITLE
Remove device assert in Gradscaler

### LINKED ISSUE #119358: enable gradscaler for other device
### 🚀 The feature, motivation and pitch

https://github.com/pytorch/pytorch/blob/main/torch/amp/grad_scaler.py#L134

Why do we need to assert here? I suggest that just provide a check. 
For third-party devices (such as privateuse1), it can also use gradscaler.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wangjiangben-hw**
Fixes #119358

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @ezyang @albanD @zou3519 @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

**Comment by ezyang** [comment]
@albanD why don't we just let them delete the assert, seems better

**Comment by albanD** [comment]
From a quick look I don't see any good reason to have that hard constraint beyond the fact that ops being used there use ops like `torch._amp_foreach_non_finite_check_and_unscale_` that are only implemented for cpu/cuda in core. So this assert is a nicer error message for end users.
I'm fine with removing it though, we can even try/catch around these calls to change the "no kernel registered for this key" error into a nicer "grad scaler required op foo is not implemented on {device}`.

Note that there are asserts for Tensors being on cpu, cuda or xla throughout the file. So the PR as it is now is not enough to get things running.

**Comment by wangjiangben-hw** [comment]
So, to delete or not to delete?  Personally, I don't care, as long as you don't jam the gradscaler at the device level.
Can you please help me with the idea? I can modify the PR based on your decision.

1.  If you don't delete it, I think the current modification is ok.
2.  If you delete it, we'll see what's appropriate next.

**Comment by wangjiangben-hw** [comment]
> From a quick look I don't see any good reason to have that hard constraint beyond the fact that ops being used there use ops like `torch._amp_foreach_non_finite_check_and_unscale_` that are only implemented for cpu/cuda in core. So this assert is a nicer error message for end users. I'm fine with removing it though, we can even try/catch around these calls to change the "no kernel registered for this key" error into a nicer "grad scaler required op foo is not implemented on {device}`.
> 
> Note that there are asserts for Tensors being on cpu, cuda or xla throughout the file. So the PR as it is now is not enough to get things running.

How do I feel about this modification? I checked, and it should be mainly these two functions that need additional error messages.

File: torch/amp/grad_scaler.py [hunk 1]
```diff
@@ -289,11 +275,15 @@ def _unscale_grads_(
...
-                    )
+                    try:
+                        torch._amp_foreach_non_finite_check_and_unscale_(
+                            grads,
+                            per_device_found_inf.get(device),
+                            per_device_inv_scale.get(device),
+                        )
+                    except Exception as e:
+                        raise NotImplementedError(f"grad scaler required op function(torch._amp_foreach_non_finite_check_and_unscale_)"
+                                                  f"is not implemented on {device}")
```
Comment (ezyang) [inline torch/amp/grad_scaler.py:286]: no, I'd prefer not to do this, suppressing the underlying exception can lead to some  pretty awful errors. E.g., if you had typoed the function name here, you'd still always hit the not implemnted case

**Review by ezyang — DISMISSED** [review]
I'm willing to take the version with only the assert removals, but if you want to be good about it, you need to update the inside of _amp_update_scale to handle unknown devices with better error message

**Comment by wangjiangben-hw** [comment]
> I'm willing to take the version with only the assert removals, but if you want to be good about it, you need to update the inside of _amp_update_scale to handle unknown devices with better error message

I agree with you. In this PR, I'll just remove the assert. @ezyang

File: torch/amp/grad_scaler.py [hunk 2]
```diff
@@ -527,7 +513,7 @@ def update(self, new_scale: Optional[Union[float, torch.Tensor]] = None) -> None
             if len(found_infs) > 1:
                 for i in range(1, len(found_infs)):
                     found_inf_combined += found_infs[i]
-
+
```
Comment (ezyang) [inline torch/amp/grad_scaler.py:516]: fix this

**Comment by wangjiangben-hw** [comment]
pull / linux-focal-py3.12-clang10 / test (default, 2, 3, linux.2xlarge) (pull_request) Failing after 88m
``` RuntimeError: Python 3.12+ not yet supported for torch.compile ```

pull / linux-focal-cuda12.1-py3.10-gcc9-sm86 / test (default, 1, 5, linux.g5.4xlarge.nvidia.gpu) (pull_request) Failing after 45m
```
test/test_transformers.py::TestSDPAFailureModesCUDA::test_flash_attention_fail_with_non_square_causal_attention_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_flash_autocast_fp32_bfloat16_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_flash_backward_failure_sm86plus_head_dim_204_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_flash_fail_fp32_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_fused_kernels_nested_broadcasting_error_cases_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_fused_kernels_nested_broadcasting_requires_grad_failure_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_attn_mask_present_kernel0_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_broadcast_kernel0_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_broadcast_kernel1_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_dim_3_kernel0_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_dim_3_kernel1_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_fused_inputs_invalid_dtype_kernel1_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_inputs_1_dimensional_inputs_kernel0_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_invalid_last_dim_stride_kernel0_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_nested_fails_on_padding_head_dim_cuda, test/test_transformers.py::TestSDPAFailureModesCUDA::test_unaligned_tensors_cuda, 
```

pull / linux-focal-cuda12.1-py3.10-gcc9-sm86 / test (default, 4, 5, linux.g5.4xlarge.nvidia.gpu) (pull_request) Failing after 72m
```
=================================== FAILURES ===================================
_______________ TraceRuleTests.test_torch_name_rule_map_updated ________________
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/dynamo/test_trace_rules.py", line 328, in test_torch_name_rule_map_updated
    self._check_set_equality(
  File "/var/lib/jenkins/workspace/test/dynamo/test_trace_rules.py", line 302, in _check_set_equality
    self.assertTrue(len(x) == 0, msg1)
  File "/opt/conda/envs/py_3.10/lib/python3.10/unittest/case.py", line 687, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true : New torch objects: {<built-in function _set_sdp_use_cudnn>, <built-in function _get_cudnn_sdp_enabled>, <built-in method _scaled_dot_product_cudnn_attention of type object at 0x7fc9f40fd380>} were not added to `trace_rules.torch_c_binding_in_graph_functions` or `test_trace_rules.ignored_c_binding_in_graph_function_names`. Refer the instruction in `torch/_dynamo/trace_rules.py` for more details.

To execute this test, run the following from the base repo dir:
     python test/dynamo/test_trace_rules.py -k test_torch_name_rule_map_updated

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
- generated xml file: /var/lib/jenkins/workspace/test/test-reports/python-pytest/dynamo.test_trace_rules/dynamo.test_trace_rules-86a5e1e7b2afa829.xml -
```

These errors doesn't seem to be related to my PR.

I see that the next 2 ci's are actually passing. Is there something wrong with the CI environment?
pull / linux-focal-cuda12.1-py3.10-gcc9 / test (default, 1, 5, linux.4xlarge.nvidia.gpu) (pull_request) Successful in 156m
pull / linux-focal-cuda12.1-py3.10-gcc9 / test (default, 4, 5, linux.4xlarge.nvidia.gpu) (pull_request) Successful in 139m

**Comment by wangjiangben-hw** [comment]
CC @ezyang @albanD

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**